[quant] Input-Weight Equalization - allow logical evaluation

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61603

Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.tensor(1)) is inserted instead. Additionally, a warning message "Must run observer before calling calculate_equalization_scale" will appear.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

angelayi committed:
[quant] Input-Weight Equalization - allow logical evaluation

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29686878).

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

supriyar commented:
What is the behavior in the case where we equalize without calibration? Do we insert the mul operator with a default scale? How can the user update the scale in this case?

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.empty(0)) is inserted instead.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -85,15 +85,17 @@ def get_input_minmax(self):
         r""" Returns the scaled min/max inputs
         """
-        if self.equalization_scale.nelement() == 0:
+        if self.equalization_scale.nelement() == 1 and self.equalization_scale == torch.tensor(1):
```
Do we need to check for `self.equalization_scale.nelement() == 1`?

angelayi replied:
I think we do, or else it will throw an error when comparing something of a different shape ex. a tensor array vs. the tensor(1).

supriyar reviewed (approved):
looks good, left a comment inline

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -66,7 +66,7 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
                                                   factory_kwargs=factory_kwargs)
 
-        self.equalization_scale = torch.empty(0)
+        self.equalization_scale = torch.tensor(1)
```
is it possible that equalization_scale is actually 1?

angelayi replied:
It could be possible that the equalization_scale is equal to 1 -- when the input column dimension is all the same number. I think the only issue with this is when scaling the next input quantization observer since we would currently use the default min/max values, but I have addressed this issue in the latest commit.

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.tensor(1)) is inserted instead. Additionally, a warning message "Must run observer before calling calculate_equalization_scale" will appear.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.tensor(1)) is inserted instead. Additionally, a warning message "Must run observer before calling calculate_equalization_scale" will appear.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

raghuramank100 reviewed (changes requested):
Testing with the input scale being 1 may not test the logic as a scale of 1 does not change the weights or activations in any way. A better test would be to use a value that is not 1.  Can we instead have the equalization scale by a fixed diagonal matrix with entries that are reproducible?.

supriyar commented:
> Testing with the input scale being 1 may not test the logic as a scale of 1 does not change the weights or activations in any way. A better test would be to use a value that is not 1. Can we instead have the equalization scale by a fixed diagonal matrix with entries that are reproducible?.

I think the goal of this PR is not to test for equalization logic, but allow the model to not throw an unexpected error if the user enables equalization but doesn't run calibration (similar to observer returning a default scale and zero-point). In this case equalization will be a no-op and the user should be able to update the scales in the state_dict on the converted model in case required before running the equalized model.

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.tensor(1)) is inserted instead. Additionally, a warning message "Must run observer before calling calculate_equalization_scale" will appear.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29686878).

angelayi committed:
Update on "[quant] Input-Weight Equalization - allow logical evaluation"


Summary: Allows us to run equalization without needing calibration. This should generate the same model structure as a calibrated and equalized model, except the equalization scales and zero points are incorrect. The mul operator and a default equalization scale (torch.tensor(1)) is inserted instead. Additionally, a warning message "Must run observer before calling calculate_equalization_scale" will appear.

The equalization scale can be updated the following way:
```python
...
equalized_quantized_model = convert_fx(prepared)
# Assume 5 is the correct equalization scale
equalized_quantized_model._buffers['x_equalization_scale0'] = torch.tensor(5)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29686878](https://our.internmc.facebook.com/intern/diff/D29686878)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29686878).
